### PR TITLE
Dataloader shutdown hangs

### PR DESCRIPTION
**Pull request by csarofeen**
So for dataloader, there's a main thread, manager thread, and worker threads. (Horrible names fyi)
Main thread does 2 things: Puts indices in a queue that gets sent to workers to process
Worker thread: reads the data and send it through the collate function and puts it to a results queue
Manager thread: Reads from the result queue, puts it on pinned memory if necessary, puts it on an output queue

Currently to shutdown dataloader processes the Main thread puts a None in the index queues, tries to empty the results queue, and then puts a None on the results queue. The manager thread / worker processes are supposed to read the None's in the queue and quit.

2 issues can cause a hang... 
1) https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L377-L378
Calling is_empty and then a get is not guaranteed to get an actual entry. Since manager thread is trying to get on this queue it could empty it in between is_empty and get and then main thread is hung on shutdown.

2) Worker threads are still running as everything goes on until they hit a None, but they may have work to do. Main thread/manager thread can quit while workers are finishing up. Main thread/manager may free queue resources and then workers will put. Since the pipes can be broken this can result in a hang on the worker threads.

I have added an extended multiprocessing queue that has a shutdown signal with guards around get/put. This will prevent attempting to communicate across the queue in any way once shutdown is triggered. It will also make sure that shutdown does not happen during a get/put. I haven't seen any more processes hang on cleanup with these modifications.

### CODE REVIEW

**Comment by SsnL** [comment]
Thanks for doing this!!! I've been trying to solve this for a while with `Queue.cancel_join_thread` but CI always fails so I gave up.

**Comment by csarofeen** [comment]
Lets see if I can get past all the checks ;-)

File: torch/multiprocessing/queue.py [hunk 1]
```diff
@@ -31,11 +32,87 @@ def __getattr__(self, name):
...
+            return False
+        with self.put_lock:
+            super(Queue, self).put(obj, block, timeout)
+            return True
+
+    def get(self, block=True, timeout=None):
+        if self.is_shutdown():
+            return None
+        with self.get_lock:
+            return super(Queue, self).get(block, timeout)
```
Comment (SsnL) [inline torch/multiprocessing/queue.py]: it's weird. why is `timeout` not triggered?

Comment (csarofeen) [reply torch/multiprocessing/queue.py]: I actually can't replicate that CI failure. When I was trying I'm not getting a hang on test_partial_workers

**Comment by csarofeen** [comment]
Sorry, there seemingly is still a bug in here somewhere... now locally I'm failing the timeout test again.

**Comment by SsnL** [comment]
`test_timeout` fails for me locally too.... but it passed on CI LOL. My feeling is that timeout does not play nicely with the locks. I'll look into this further tomorrow.

**Comment by SsnL** [comment]
@csarofeen Here are some interesting behavior I saw:
1. Moving the `super().get` call outside `with self.get_lock` passes the test, i.e., with change 
   ```py
    def get(self, block=True, timeout=None):
        if self.is_shutdown():
            return None
        return super(Queue, self).get(block, timeout)
   ````
2. With the above change, directly using `worker_result_queue` as `data_queue` when `timeout > 0` fails the test, i.e., with above change and 
    ```py
        -  if self.pin_memory or self.timeout > 0:
        +  if self.pin_memory:
                self.data_queue = queue.Queue()
                ..... # init worker_manager_thread thread
            else:
                self.data_queue = self.worker_result_queue
    ```
    Directly using `worker_result_queue` should be fine when `timeout > 0` because it is now a `mp.Queue` rather than a `SimpleQueue`.

**Comment by SsnL** [comment]
I got myself a CI docker and `test_timeout` fails in docker lol.

**Comment by csarofeen** [comment]
I'm at a real loss here, if you have any ideas I'm willing to run with them. get/put breaking in test_timeout is really strange to me.

**Comment by SsnL** [comment]
@csarofeen I'm looking into this as we speak. (really want this in 0.4.1

**Comment by SsnL** [comment]
@csarofeen I know why test_timeout fails locally now:
1. Timeout is triggered. (Some) worker is still waiting to get, holding the `get_lock`.
2. In `__del__`, `for q in self.index_queues: q.shutdown()` tries to acquire the lock, causing the hang.

So this locking mechanism seems quite unsafe if main process has some error. :O

**Comment by csarofeen** [comment]
That's why part of the shutdown sequence does a put on None before it tries to grab the get lock. The none should allow the get's to finish.

File: torch/multiprocessing/queue.py [hunk 2]
```diff
@@ -31,11 +32,91 @@ def __getattr__(self, name):
...
+
+    def __getstate__(self):
+        return (super(Queue, self).__getstate__(),
+                (self.sig_shutdown, self.put_lock, self.get_lock))
+
+    def shutdown(self):
+        if self.is_shutdown():
+            return
+        with self.sig_shutdown.get_lock():
+            with self.put_lock:
```
Comment (csarofeen) [inline torch/multiprocessing/queue.py]: Grab the put lock, put a None to try free the get lock, then signal shutdown. This makes sure no one is getting or putting when we try to shutdown, preventing a race condition on shutdown.

**Comment by SsnL** [comment]
@csarofeen You are right. It should prevent that case indeed. I'm not understanding this either....

**Comment by SsnL** [comment]
@csarofeen I don't understand why, but this has some problem when the process raises error. I tried this patch on the new `test_manager_unclean_exit` from #9655  and it fails the case that main process raises an error. 

Additionally, I don't think we should arbitrarily add a lock mechanism to the `mp.Queue` (it should only be augmented with the pickling mechanism on tensors). 

Given these, how do you think about we use #9655 for 0.4.1? That one passes the test and seems fine.

**Comment by csarofeen** [comment]
Closing in favor of https://github.com/pytorch/pytorch/pull/9655
